### PR TITLE
Adds deflate compression to MsgState.

### PR DESCRIPTION
**Pull request by gradientvera**
Related #889
GameStates above a certain size threshold (256 bytes right now) will be compressed using the deflate algorithm.
I've tested this extensively, and it seems to produce very good results. Here is an example:
image
Initial gamestate being compressed from 1.38 MB down to 73.22 KB

Paging @PJB3005 for review

### CODE REVIEW

File: Robust.Shared/Network/Messages/MsgState.cs [hunk 1]
```diff
@@ -26,27 +31,66 @@ public class MsgState : NetMessage
...
+            var uncompressedLength = buffer.ReadVariableInt32();
+            var compressedLength = buffer.ReadVariableInt32();
+            MemoryStream finalStream;
+
+            // State is compressed.
+            if (compressedLength > 0)
+            {
+                var stream = buffer.ReadAlignedMemory(compressedLength);
+                using var decompressStream = new DeflateStream(stream, CompressionMode.Decompress);
+                var decompressedStream = new MemoryStream(decompressStream.CopyToArray(), false);
```
Comment (PJB3005) [inline Robust.Shared/Network/Messages/MsgState.cs]: When you do `.CopyToArray()`, the implementation makes a `MemoryStream` which has to gradually grow its internal buffer. You should probably have `decompressedStream` be initialized to an empty stream (of the correct size) and then decompress into it with `decompressStream`. Then just `.Position = 0` to reset it and you can pass it down to the deserializer.

File: Robust.Shared/Network/Messages/MsgState.cs [hunk 2]
```diff
@@ -26,27 +31,66 @@ public class MsgState : NetMessage
...
+            State.PayloadSize = uncompressedLength;
         }
 
         public override void WriteToBuffer(NetOutgoingMessage buffer)
         {
             var serializer = IoCManager.Resolve<IRobustSerializer>();
-            using (var stateStream = new MemoryStream())
+            MemoryStream finalStream;
+            var stateStream = new MemoryStream();
+            DebugTools.Assert(stateStream.Length <= Int32.MaxValue);
```
Comment (PJB3005) [inline Robust.Shared/Network/Messages/MsgState.cs]: I would be pretty concerned aswell if a stream I just created and should thus be empty had a length that big, yeah.

File: Robust.Shared/Network/Messages/MsgState.cs (hunk 1 above)
Comment (gradientvera) [reply Robust.Shared/Network/Messages/MsgState.cs]: Should be fixed now, I think.

File: Robust.Shared/Network/Messages/MsgState.cs (hunk 2 above)
Comment (gradientvera) [reply Robust.Shared/Network/Messages/MsgState.cs]: Oops

**Pull request merged**